internal/clients/telegram/client.go

```diff
@@ -264,231 +309,171 @@ However it is not as simple as that. After an update is processed it could chang
 
-			return
+		if handle, ok := upd.StateID(); ok {
+			futureState = c.borrowState(handle)
+		}
+
+		futureUserData := borrowonce.NewImmediateFuture[state.UserSharedData](state.UserSharedData{
+			GithubAPIKey: option.None[string](),
+		})
+
+		if handle, ok := upd.UserID(); ok {
+			futureUserData = c.borrowUserData(handle)
 		}
 
 		stateCh <- updateWithState{
-			update: upd,
-			state:  state,
+			update:   upd,
+			state:    futureState,
+			userData: futureUserData,
 		}
 	}
 
-	c.wg.Done()
-	close(stateCh)
+	shutdown()
 }
 
 /*
-TODO(takeState, releaseState): Once takeState gives out an instance of state for a conversation it should not be able to
-give out another instance. If there are 2 things that hold on to state to the same conversation they could try to update
-it (save it to state storage) which would result in one state being lost and the final state is whatever was saved last.
-They could also be operating on invalid state.
-
-Consider 2 updates: /quiz and then /start. And lets say /quiz makes the bot send a question and the next message (in
-this case /start) would be the answer. If we process /quiz and /start at the same time /start update would have no idea
-that it should be handled as an answer to /quiz, and the user will get 2 messages: the first one is the quiz question
-and the second one is the greeting message from /start. Then the user tries to answer the quiz and nothing happens
-because /start saved the state last and overwrote what /quiz did.
-
-TODO: If the state for a conversation was given out and not released via releaseState() this should fail or block.
+borrowState returns a Future to access the latest value of the state. If no state is in the storage then assigns it to
+Root.
 */
-func (c *Client) takeState(u update.Update) (state.Handler, error) {
-	handle, isSome := u.StateID().Unwrap()
-	if isSome {
-		if state, found := c.conversationStates[handle]; found {
-			err := state.SetTemplate(c.template)
-			if err != nil {
-				log.Printf("While fetching state from storage for handle %q: %s", handle, err)
-			} else {
-				log.Printf("Lent state %T", state)
-
-				return state, nil
-			}
-		}
+func (c *Client) borrowState(handle string) *borrowonce.Future[state.State] {
+	if future, exists := c.conversationStateStore.Borrow(handle); exists {
+		return future
 	}
 
-	state := &state.Root{}
+	c.conversationStateStore.Set(handle, &state.RootState{})
 
-	if err := state.SetTemplate(c.template); err != nil {
-		return state, fmt.Errorf("while fetching state from storage for the default state: %w", err)
+	if future, exists := c.conversationStateStore.Borrow(handle); exists {
+		return future
 	}
 
-	log.Printf("Lent state %T", state)
-
-	return state, nil
+	// Implementation of borrownonce.Store guarantees that after a value is Set() it is available to borrow
+	panic("conversation store did not lend a value after it was set explicitly")
```

Comment:
This panic message should include a function name and or generic types on the storage.